### PR TITLE
Use grig

### PR DESCRIPTION
**Pull request by thomasjwebb**
There is some more cleanup I think I could do to this, like deleting unused code although some of it might be useful later. I started over on the fft side and ported over audacious' visualization code into grig and then modified this to use that. But still using our windowing function, the peak history thing, etc. It's pretty performant. In my tests at least at slowest it ran in 1/500ths of a second. There are things I could do to reduce the cpu and memory footprint though if it needs to be faster for whatever reason. The main two things that we were actually doing wrong before was:

- Not converting the audio data correctly. This was a dumb obvious thing I somehow missed but we were just reading halves of the 16-bit values and using them as though they were 16-bit values.
- Some subtle bad math (dividing by 2 when not needed, etc.) that caused it to effectively run at half speed.

I could probably make the old code work knowing this but in porting the audacious code I set up unit tests to check every step of the way that my port produces exactly the same result as the C++ original.

### CODE REVIEW

**Comment by thomasjwebb** [comment]
For reference, the changes I made to grig that made this possible:
https://gitlab.com/haxe-grig/grig.audio/-/merge_requests/2/diffs

Not merging over there until I fix the unit testing which appears to just be that I need to switch to a new docker image

File: src/funkin/vis/dsp/SpectralAnalyzer.hx [hunk 1]
```diff
@@ -1,476 +1,125 @@
...
+		var signal = segment.toInterleaved(audioSource.buffer.bitsPerSample);
+
+		if (numChannels > 1) {
+			var mixed = new Array<Float>();
+			mixed.resize(Std.int(signal.length / numChannels));
+			for (i in 0...mixed.length) {
+				mixed[i] = 0.0;
+				for (c in 0...numChannels) {
+					mixed[i] += 0.7 * signal[i*numChannels+c];
+                    break;
```
Comment (NeeEoo) [inline src/funkin/vis/dsp/SpectralAnalyzer.hx:60]: Whats with this break?

Comment (thomasjwebb) [reply src/funkin/vis/dsp/SpectralAnalyzer.hx:60]: That was left in by mistake! I'll fix that.

Comment (thomasjwebb) [reply src/funkin/vis/dsp/SpectralAnalyzer.hx]: Okay I fixed that. I meant to just temporarily only use the left channel while debugging the weird data tear issue that was caused by `currentFrame` sometimes being an odd number. And I can confirm removing the break didn't break anything.

**Review by mcagabe19 — CHANGES_REQUESTED** [review]
lime's audiosource is returning null at html5
image
image-1

**Comment by thomasjwebb** [comment]
> lime's audiosource is returning null at html5 (can due removal of `funkin.vis._internal.html5.AnalyzerNode`?) 

Ah okay I'm not surprised that this is brittle since we're relying on internals of flixel being at a certain place. Since `audioSource` comes from:

```haxe
FlxG.sound.music._channel.__source
```

I'm not familiar with `funkin.vis._internal.html5.AnalyzerNode` does that provide a shim to let us access the playing music channel?

**Comment by thomasjwebb** [comment]
I also wonder if it's just that the sound hasn't finished loading yet...

**Comment by thomasjwebb** [comment]
Okay yeah I see the problem. We are relying too much on specifics of the internals. I'll make a wrapper class to work around this. Non-html5 targets set `data` to something, html5 sets `src` to a different thing.

**Comment by thomasjwebb** [comment]
Ah okay it looks like flixel and lime are basically using howlerjs under the hood. I'll post about this in discord to make sure there aren't cases I'm not considering.

**Comment by KarimAkra** [comment]
funkin.vis was using `funkin.vis._internal.html5.AnalyzerNode`'s `getFloatFrequencyData` method to obtain the audio's frequency/amplitude
it uses `js.html.audio.AnalyserNode` which is webapi as far as i know and not FFT
this method seemed more accurate at visualizing the audio and more optimized
maybe it should be added back?

**Comment by ninjamuffin99** [comment]
correct yeah for HTML5 i used a lil wrapper over the web audio api stuff to use the browsers native FFT solution, which I've found to be suuuuuper performant, as its already been optimized to hell and back by google + mozilla

**Comment by thomasjwebb** [comment]
I think you're talking about code I don't have access to. I think the example in this repo is just a test for the code here, not necessarily representative of what the game's doing. Yeah webapi is actually pretty good so it's best to use that when you can, since it's using optimized native code under the hood, then fall back to using the in-haxe fft when that's not available.

**Comment by thomasjwebb** [comment]
I can make this work in html5 too if need be. What I'd have to do is have a class that does some trickery accessing the internals of howler.js to get at the raw audio stream. But if this is only used on non-html5 targets, that shouldn't be necessary, right?

**Comment by thomasjwebb** [comment]
Aah! I misunderstood! Yes you were calling that in this function and I inadvertently removed that. I'll add that back in.

**Comment by thomasjwebb** [comment]
Okay I have a question then. It seems I don't have these implemented anymore since I replaced this file:

```haxe
public var fftN(default, set):Int = 4096;
public var minDb(default, set):Float = -70;
public var maxDb(default, set):Float = -20;
```

Do you need to set these parameters after construction? If so I need to make sure they're implemented here too.

**Comment by thomasjwebb** [comment]
Okay I fixed that. It's not super elegant but now it works the way it did before on html5 target but still also works in hxcpp targets. There are differences in how it looks of course - the big thing is the frequency range is different.

**Review by ninjamuffin99 — APPROVED** [review]

**Comment by ninjamuffin99** [comment]
thank you thomas you ar a hero !

**Pull request merged**